stage0/src/sev.rs

```diff
@@ -28,14 +29,106 @@ use oak_sev_guest::{
 use spinning_top::Spinlock;
 use x86_64::{
     instructions::tlb,
-    registers::control::Cr3,
+    registers::{control::Cr3, model_specific::Msr},
     structures::paging::{
         frame::PhysFrameRange, page::AddressNotAligned, Page, PageSize, PageTable, PageTableFlags,
         PhysFrame, Size1GiB, Size2MiB, Size4KiB,
     },
     PhysAddr, VirtAddr,
 };
 
+/// The cache memory type used with MTRR.  We only use Write-Protect mode which the Linux kernel
+/// expects to be enabled by the firmware in order to enable SEV.
+#[repr(u8)]
+#[allow(dead_code)] // Remove if this is ever ported to a public crate.
+pub enum MemoryType {
+    UC = 0, // Uncacheable.
+    WC = 1, // Write-Combining.
+    WT = 4, // Writethrough	Reads.
+    WP = 5, // Write-Protect.
+    WB = 6, // Writeback.
+}
+
+impl MemoryType {
+    pub fn from_u8(value: u8) -> MemoryType {
```

Comment:
Rust has a `TryFrom` trait for just these kinds of operations!

Something like this:
```
impl TryFrom<u8> for MemoryType {
  type Error = &'static str; // or a better error type
  fn try_from(value: u8) -> Result<Self, &'static str> {
   // return Ok(MemoryType::something) or Err("invalid memory type") here
}
}
```

Reply:
Honestly, this is a fun CL, as I'm learning about Rust during this review.  Sorry for being so green.

The second approach would be the way to go, but the strum crate depends on std, and we compile with no_std.  The first approach seems more like the code in oak_sev_guest/src/msr.rs.  I happen to dislike the duplication of constants so I hacked a way to avoid them, but the additional complexity may not be worth it.

WDYT?  I don't believe that short of writing a custom macro, I can avoid the match expression in fn try_from.  This seems like a big enough hole in running without std that maybe the macro should be written?  It seems silly to have overly complex match expressions in our try_from functions for enums.  A custom derive macro might be fun to write.

Reply:
IIRC strum should work with no std support by not enabling default features (though I'm on my phone now and cannot check)

Reply:
That worked!  I should have noticed that from the other places that use strum.